src/looprestoration.rs

```diff
@@ -404,64 +408,49 @@ unsafe fn wiener_rust<BD: BitDepth>(
         edges,
     );
 
-    let mut hor: [uint16_t; 27300] = [0; 27300];
-    let mut hor_ptr: *mut uint16_t = hor.as_mut_ptr();
-    let filter: *const [int16_t; 8] = ((*params).filter.0).as_ptr();
-    let bitdepth = bd.bitdepth().as_::<libc::c_int>();
-    let round_bits_h = 3 as libc::c_int + (bitdepth == 12) as libc::c_int * 2;
-    let rounding_off_h = (1 as libc::c_int) << round_bits_h - 1;
-    let clip_limit = (1 as libc::c_int) << bitdepth + 1 + 7 - round_bits_h;
-    let mut j = 0;
-    while j < h + 6 {
-        let mut i = 0;
-        while i < w {
-            let mut sum = (1 as libc::c_int) << bitdepth + 6;
+    // Values stored between horizontal and vertical filtering don't
+    // fit in a uint8_t.
+    let mut hor = [0u16; 70 /*(64 + 3 + 3)*/ * REST_UNIT_STRIDE];
 
-            if BD::BITDEPTH == 8 {
-                sum += (*tmp_ptr.offset((i + 3) as isize)).as_::<libc::c_int>() * 128;
+    let filter = &params.filter.0;
+    let bitdepth = bd.bitdepth().as_::<libc::c_int>();
+    let round_bits_h = 3 + (bitdepth == 12) as libc::c_int * 2;
+    let rounding_off_h = 1 << round_bits_h - 1;
+    let clip_limit = 1 << bitdepth + 1 + 7 - round_bits_h;
+    let chunks = tmp
+        .chunks_exact(REST_UNIT_STRIDE)
+        .zip(hor.chunks_exact_mut(REST_UNIT_STRIDE));
+    for (tmp, hor) in chunks.take((h + 6) as usize) {
+        for i in 0..w as usize {
+            let mut sum = 1 << bitdepth + 6;
+
+            if BD::BPC == BPC::BPC8 {
+                sum += tmp[i + 3].as_::<libc::c_int>() * 128;
             }
 
-            let mut k = 0;
-            while k < 7 {
-                sum += (*tmp_ptr.offset((i + k) as isize)).as_::<libc::c_int>()
-                    * (*filter.offset(0))[k as usize] as libc::c_int;
-                k += 1;
+            for k in 0..7 {
+                sum += tmp[i + k].as_::<libc::c_int>() * filter[0][k] as libc::c_int;
             }
```

Comment:
```suggestion
            for (&tmp, &filter) in std::iter::zip(&tmp[i..][..7], &filter[0][..7]) {
                sum += tmp.into() * filter as libc::c_int;
            }
```

Bounds checking elision.

Reply:
We can save this for later, though.